### ISSUE TITLE
[docs] Potential inconsistency between documentation and code for the compiler

### ISSUE THREAD

**Issue by tan-wei** [ethereum-optimism/optimism#1091]
**Describe the bug**
In the [document](https://community.optimism.io/docs/protocol/protocol.html#execution-contracts), it says that
```
* The CALLER, CALL, and REVERT opcodes are also disallowed, except in the special case that they appear as part of one of the following strings of bytecode:

    CALLER PUSH1 0x00 SWAP1 GAS CALL PC PUSH1 0x0E ADD JUMPI RETURNDATASIZE PUSH1 0x00 DUP1 RETURNDATACOPY RETURNDATASIZE PUSH1 0x00 REVERT JUMPDEST RETURNDATASIZE PUSH1 0x01 EQ ISZERO PC PUSH1 0x0a ADD JUMPI PUSH1 0x01 PUSH1 0x00 RETURN JUMPDEST
```
However, we can see the code is implemented [here](https://github.com/ethereum-optimism/solidity/blob/df005f39493525b43f1153dff8da5910a2b83e34/libyul/backends/evm/EVMDialect.cpp#L119), but some differences, for example: `0x0E` vs `_assembly.appendConstant(29)`.

Any inconsistence here? Or my misunderstanding about the document or code? Thanks very much.

**Comment by smartcontracts** [comment]
I've asked @ben-chain to answer this question as he has the most context here.

**Comment by ben-chain** [comment]
Hey @tan-wei , this is a great question--the answer has to do with some of the limitations of yul/solidity at the time the compiler was written.  In particular, what you see in `CompilerContext.cpp` is us "sidestepping the optimizer" -- it is a bytestring of the same number of bytes that we want, but un-optimizable. The optimizer really did not like the `JUMPDEST`s in the documented "safe bytestring" and would toss it out.  So what we did was chose a value that the optimizer would never think it can improve upon.  Then, as a final step in the compilation process, we actually go in and [find/replace with the "real" string,](https://github.com/ethereum-optimism/solidity/compare/7338295..develop-0.7#diff-0f19df92d55a7d7fe38fdf4be1e14b364671c4934853d13a33b31aedf3f917bdR1168-R1280) once the optimizer has already been applied.  So the output of the compiler is ultimately consistent with the documentation at the end of compilation, but wrong in the middle, so that the optimizer doesn't toss it out. 😅

This is why we have worked with the solidity team to get the `verbatim` builtin into yul.  This has been [included in the latest release](https://blog.soliditylang.org/2021/06/10/solidity-0.8.5-release-announcement/), and is documented [here](https://docs.soliditylang.org/en/v0.8.5/yul.html#verbatim).  From the docs: "It also allows you to create bytecode sequences that will not be modified by the optimizer."  This is exactly what will allow us to get rid of the confusing logic you've noticed. :)
